strawberry/dataloader.py

```diff
@@ -0,0 +1,125 @@
+import dataclasses
+from asyncio import create_task, get_event_loop
+from asyncio.futures import Future
+from dataclasses import dataclass
+from typing import Any, Awaitable, Callable, Dict, Generic, List, Optional, TypeVar
+
+from .exceptions import WrongNumberOfResultsReturned
+
+
+T = TypeVar("T")
+K = TypeVar("K")
+
+
+@dataclass
+class LoaderTask:
+    key: Any
+    future: Future
+
+
+@dataclass
+class Batch:
+    tasks: List[LoaderTask] = dataclasses.field(default_factory=list)
+    dispatched: bool = False
+
+    def add_task(self, key: Any, future: Future):
+        task = LoaderTask(key, future)
+        self.tasks.append(task)
+
+    def __len__(self) -> int:
+        return len(self.tasks)
+
+
+class DataLoader(Generic[T, K]):
+    queue: List[LoaderTask] = []
+    batch: Optional[Batch] = None
+    cache: bool = False
+    cache_map: Dict[K, Future]
+
+    def __init__(
+        self,
+        load_fn: Callable,
```

Comment:
is it possible the type of load_fn more specific? ie it's a callable that takes an array of keys `K` and returns an array of optional results `R`, then you could type the cache and the return value of `load`.

Reply:
Actually, we already have K, but we're not using T from the generic above.

Reply:
Done! Added some tests too 😊